src/core/ComboBox/ComboBox.tsx

```diff
@@ -117,252 +122,312 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+  React.useEffect(() => {
+    if (selectedIndex != undefined && selectedIndex >= 0) {
+      const value = options[selectedIndex]?.value;
+      if (value === valueProp) {
+        return;
+      }
+      onChangeProp.current?.(value);
     }
+  }, [options, selectedIndex, valueProp]);
+
+  // Memoize the default (not selected/focused) version of menu items
+  const memoizedItems = React.useMemo(() => {
+    return options.map((option, index) => {
+      const { __originalIndex, ...rest } = option;
 
-    const _filteredOptions =
-      filterFunction?.(options, inputValue) ??
-      options.filter((option) =>
-        option.label.toLowerCase().includes(inputValue?.trim().toLowerCase()),
+      const customItem = itemRenderer
+        ? itemRenderer(option, {
+            isFocused: false,
+            isSelected: false,
+            index: __originalIndex,
+            id: option.id,
+          })
+        : null;
+
+      return customItem ? (
+        React.cloneElement(customItem, {
+          'data-iui-index': __originalIndex,
+          onClick: (e: unknown) => {
+            dispatch(['select', __originalIndex as number]);
+            customItem.props.onClick?.(e);
+          },
+        })
+      ) : (
+        <ComboBoxMenuItem
+          key={(__originalIndex as number) ?? index}
+          data-iui-index={option.__originalIndex}
+          onClick={() => dispatch(['select', __originalIndex as number])}
+          {...rest}
+        >
+          {option.label}
+        </ComboBoxMenuItem>
       );
-    setFilteredOptions(_filteredOptions);
-
-    setFocusedIndex((previouslyFocusedIndex) => {
-      if (_filteredOptions.includes(options[previouslyFocusedIndex])) {
-        return previouslyFocusedIndex;
-      } else if (
-        _filteredOptions.find(({ value }) => value === selectedValue)
-      ) {
-        return options.findIndex(({ value }) => value === selectedValue);
-      } else {
-        return -1; // reset focus if previously focused or selected value is not in filtered list
-      }
     });
-  }, [inputValue, options, selectedValue, isOpen, filterFunction]);
-
-  const onKeyDown = React.useCallback(
-    (event: React.KeyboardEvent) => {
-      const focusableOptions = getFocusableElements(menuRef.current);
-      const focusedIndexInFilteredList = focusableOptions.findIndex(
-        ({ id }) =>
-          id === inputRef.current?.getAttribute('aria-activedescendant'),
+  }, [itemRenderer, options]);
+
+  const getMenuItem = React.useCallback(
+    (option: ComboBoxOption<T>) => {
+      const { __originalIndex, ...rest } = option;
+
+      const customItem = itemRenderer
+        ? itemRenderer(option, {
+            isFocused: focusedIndex === __originalIndex,
+            isSelected: selectedIndex === __originalIndex,
+            index: __originalIndex,
+            id: option.id,
+          })
+        : null;
+
+      return selectedIndex === __originalIndex ||
```

Comment:
Can you change to normal `if` condition. For me it is hard to read several ternary conditionals in one place.

Reply:
Removed nested ternary, take a look